pkg/ipam/allocator/multipool/pool_allocator.go

```diff
@@ -105,20 +106,127 @@ func (p *PoolAllocator) RestoreFinished() {
 	p.mutex.Unlock()
 }
 
+func (p *PoolAllocator) addPool(poolName string, ipv4CIDRs []string, ipv4MaskSize int, ipv6CIDRs []string, ipv6MaskSize int) error {
+	v4, err := cidralloc.NewCIDRSets(false, ipv4CIDRs, ipv4MaskSize)
+	if err != nil {
+		return err
+	}
+
+	v6, err := cidralloc.NewCIDRSets(true, ipv6CIDRs, ipv6MaskSize)
+	if err != nil {
+		return err
+	}
+
+	p.pools[poolName] = cidrPool{
+		v4:         v4,
+		v6:         v6,
+		v4MaskSize: ipv4MaskSize,
+		v6MaskSize: ipv6MaskSize,
+	}
+
+	return nil
+}
+
 func (p *PoolAllocator) AddPool(poolName string, ipv4CIDRs []string, ipv4MaskSize int, ipv6CIDRs []string, ipv6MaskSize int) error {
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 
-	if _, ok := p.pools[poolName]; ok {
+	if _, exists := p.pools[poolName]; exists {
 		return fmt.Errorf("pool %q already exists", poolName)
 	}
+	return p.addPool(poolName, ipv4CIDRs, ipv4MaskSize, ipv6CIDRs, ipv6MaskSize)
+}
 
-	v4, err := cidralloc.NewCIDRSets(false, ipv4CIDRs, ipv4MaskSize)
+func (p *PoolAllocator) updateCIDRSets(isV6 bool, cidrSets []cidralloc.CIDRAllocator, newCIDRs []string, maskSize int) ([]cidralloc.CIDRAllocator, error) {
+	var newCIDRSets []cidralloc.CIDRAllocator
+	var alloc []string
+	var err error
+
+	// allocate new CIDR set for each CIDR not yet in the pool
+	for _, cidr := range newCIDRs {
+		prefix, parseErr := netip.ParsePrefix(cidr)
+		if parseErr != nil {
+			err = errors.Join(err, fmt.Errorf("invalid CIDR: %w", parseErr))
+			continue
+		}
+		if !hasCIDR(cidrSets, prefix) {
+			alloc = append(alloc, cidr)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+	if len(alloc) > 0 {
+		newCIDRSets, err = cidralloc.NewCIDRSets(isV6, alloc, maskSize)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// delete CIDR set for CIDRs not present in the new CIDRs
+	for i, oldCIDR := range cidrSets {
+		exists := false
+		for _, cidrString := range newCIDRs {
+			_, cidr, parseErr := net.ParseCIDR(cidrString)
+			if parseErr != nil {
+				err = errors.Join(err, fmt.Errorf("invalid CIDR: %w", parseErr))
+				continue
+			}
+			if oldCIDR.IsClusterCIDR(cidr) {
+				exists = true
+				break
+			}
+		}
+		if !exists {
+			cidrSets[i] = nil
+			cidrSets = slices.Delete(cidrSets, i, i+1)
+
+			prefix := oldCIDR.Prefix()
+
+			for node, pools := range p.nodes {
+				for pool, allocatedCIDRSets := range pools {
+					if isV6 {
+						if _, ok := allocatedCIDRSets.v6[prefix]; ok {
```

Comment:
If we delete prefixes from nodes, then we assume that they are still in use. We should emit a warning here